design-proposals/klip-17-sql-union.md

```diff
@@ -0,0 +1,193 @@
+
+## What is not in scope
+
+* Introduction of `UNION` operator: which removes duplicates from the result set.
+We may choose to introduce this later, but is not required in any MVP.
+* Introduction of `UNION ALL` support on tables. though it may not be much work to add this later. 
+Support for tables is not required to remove `INSERT INTO` and unblock the other KLIPs, 
+hence it has been de-scoped.
+* And changes to `INSERT VALUES` functionality: this KLIP does not address any issues with `INSERT VALUES`. 
+These will be covered by [KLIP-21](klip-21_correct_insert_values_semantics.md).
+* Type coercion: the SQL standard only requires the matching columns from each `SELECT` in the union
+to have 'similar data types'. Such type coercion is not required for an MVP and may be added later. 
+
+## Value/Return
+
+This KLIP is mainly about enabling other improvements, (e.g. KLIP-18 through 21), though it will also
+simplify the KSQL conceptual model, reducing the cognitive load on users. 
+
+## Public APIS
+
+* Remove of `INSERT INTO` support from the supported SQL syntax.
+* Add `UNION ALL` support for streams.
+
+## Design
+
+See [UNION docs on w3schools](https://www.w3schools.com/sql/sql_union.asp) for examples of how `UNION ALL` works.
+
+Under the hood this PR will require support for sub-queries to be added to our query planning pipeline.
+Each `SELECT` in the union will be added to the logical plan as a sub-query, and translated into a suitable
+physical plan, and ultimately building built into a single KS topology.
+
+The CLI will be updated to detect `INSERT INTO` statements and inform the user they are no longer supported, 
+and what to do instead, linking to more documentation. This can be done through a general pattern that can 
+be used for other removed statement types.
+
+## Test plan
+
+Suitable `QueryTranslationTest` cases will be added to cover the use of `UNION ALL`. 
+At a minimum this should replicate the QTT tests that exist for `INSERT INTO`.
+
+If [KLIP-6: Execution Plans](klip-6-execution-plans.md) has not been merged then any existing 
+`INSERT ALL` statements in the command topic will result in an error.  If it has been merged, 
+the query may execute, which would be bad, unless we add code to explicitly disable this. 
+Hence testing in this area is crucial. 
+
+## Documentation Updates
+
+The removal of `INSERT INTO` will obviously necessitate many areas of documentation being updated. 
+Luckily `INSERT INTO` is not heavily used in demos or examples.
+
+A new section will be added to the syntax reference to cover the use of `UNION ALL`. Not much beyond 
+that would be needed in our docs as this is a standard SQL operator, and hence well documented. 
+
+The release notes will include details of how to migrate `INSERT INTO` statements to use `UNION ALL`.
+
+# Compatibility Implications
+
+Removal of `INSERT ALL` is clearly a breaking change. However, we _must_ remove it if we are to unlock
+the other planned work. 
```

Comment:
`INSERT ALL` ? Should be `INSERT INTO`?

What is unclear in general about this KLIP is, why not just add `UNION ALL` for streams, and keep `INSERT INTO` as-is? Why remove `INSERT INTO`? What other work is planned to use it?

Reply:
> INSERT ALL ? Should be INSERT INTO?

updated!

> What is unclear in general about this KLIP is, why not just add UNION ALL for streams, and keep INSERT INTO as-is? Why remove INSERT INTO? What other work is planned to use it?

I thought the document was being quite clear on this. It has a section on it:

```
The later drawbacks are stopping other improvements to KSQL. Including, though not limited to:

* [KLIP-18][18]: Distributed Metastore: which requires the (1 - 0..1)
relationship between data source and persistent query that this KLIP will restore.
See the KLIP-17 section in KLIP-18 for details of why this is needed.
* [KLIP-19][19]: Introduce Materialized Views: which looks to clarify the
semantic difference between a view and a table/stream and their associated semantics. 
This KLIP is related because it removes `INSERT INTO` which inserts into an existing stream.
* [KLIP-20][20]: Remove `TERMINATE` statement: which requires the (1 - 0..1)
relationship between data source and persistent query that this KLIP will restore.
* [KLIP-21][21]: Correct 'INSERT VALUES' semantics: which depends on [KLIP-19][19].
```